[AIRFLOW-2425] Maintain lineage and make available in other systems

bolkedebruin opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2425
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.


### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

Lineage 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:


### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- X ] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.


### Code Quality
- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`


Will squash when tests pass and comments are integrated.

cc: @criccomini @jgao54 @Fokko 

<img width="771" alt="screen shot 2018-05-05 at 23 54 55" src="https://user-images.githubusercontent.com/4282712/39667889-0eaa6898-50c0-11e8-84b9-8eb90f98294f.png">

Fokko commented on airflow/lineage/datasets.py:
```diff
@@ -0,0 +1,140 @@
+    )
+
+
+def map_type(name):
```
Move this as a static method inside the `DataSet` class

bolkedebruin replied:
will do

Fokko commented on airflow/lineage/datasets.py:
```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
```
Can we split the different files to different files? I maybe move these files in the dataset module.

bolkedebruin replied:
I think that is a bit over kill. At the moment datasets are designed to be light weight so it doesnt make much sense to split it out right now.

Fokko commented on tests/lineage/test_lineage.py:
```diff
@@ -0,0 +1,118 @@
+        ctx1 = {"ti": TI(task=op1, execution_date=DEFAULT_DATE)}
+        ctx2 = {"ti": TI(task=op2, execution_date=DEFAULT_DATE)}
+        ctx3 = {"ti": TI(task=op3, execution_date=DEFAULT_DATE)}
+        # ctx4 = {"ti": TI(task=op4, execution_date=DEFAULT_DATE)}
```
Why is this one commented out?

bolkedebruin replied:
No test required I can remove it

Fokko commented on docs/lineage.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. code:: python
+
+    """
```
Missing closing quotes of the comments? Maybe move the comments outside of the code block

Fokko commented on docs/lineage.rst:
```diff
@@ -0,0 +1,87 @@
+Lineage
```
Maybe mention it is still in alpha?

bolkedebruin replied:
It will be in master not in a release yet

Fokko commented on airflow/lineage/backend/atlas/__init__.py:
```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
```
For me it makes more sense that we have an interface that is extended by an implementation of a lineage backend.

bolkedebruin replied:
You mean using an abstract class? It looks like overkill right now (only one function is called), but can do

Fokko commented on airflow/operators/bash_operator.py:
```diff
@@ -89,6 +89,7 @@ def execute(self, context):
         bash_command = ('export {}={}; '.format(AIRFLOW_HOME_VAR, airflow_home_value) +
                         'export {}={}; '.format(PYTHONPATH_VAR, pythonpath_value) +
                         self.bash_command)
+        self.lineage_data = bash_command
```
Can't we also do this using a decorator, like we do with the `{pre,post}_execute`.

bolkedebruin replied:
How to do that? It is an arbitrary assignment, so how do I know which variable to catch?

msumit commented on airflow/operators/email_operator.py:
```diff
@@ -35,7 +35,7 @@ class EmailOperator(BaseOperator):
     :param files: file names to attach in email
     :type files: list
-    :param cc: list of recipients to be added in CC field
+    :param cc: list of recipients to be added in CC fieldn
```
typo `fieldn`?

bolkedebruin replied:
Good catch. Should not have changed this file

bolkedebruin committed:
[AIRFLOW-2425] Add lineage support

Add lineage support by having inlets and oulets that
are made available to dependent upstream or downstream
tasks.

If configured to do so can send lineage data to a
backend. Apache Atlas is supported out of the box.

Fokko reviewed (approved):
LGTM
